pkg/flavors/flavors.go

```diff
@@ -0,0 +1,80 @@
+// Copyright 2020 Tetrate
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package flavors
+
+import (
+	"bytes"
+	"fmt"
+	"text/template"
+)
+
+// FlavorConfigTemplate - interface to individual flavors.
+type FlavorConfigTemplate interface {
+	CheckParams(params map[string]string) (interface{}, error)
+	GetTemplate() string
+}
+
+// Main repo for templates.
+type templateStore struct {
+	// This is a map indexed by flavor pointing to the individual
+	// implementaions of each flavor.
+	templates map[string]FlavorConfigTemplate
+}
+
+var store = templateStore{templates: make(map[string]FlavorConfigTemplate)}
+
+// AddTemplate - function is used by individual flavors (like postgres)
+// to add the flavor to main repo.
+func AddTemplate(flavor string, configTemplate FlavorConfigTemplate) {
+	store.templates[flavor] = configTemplate
+}
+
+// GetTemplate - function returns FlavorConfigTemplate structure associated
+// with flavor.
+func GetTemplate(flavor string) (FlavorConfigTemplate, error) {
```

Comment:
I think we need more than flavor to infer a template, flavor:config-template isn't 1:1. For example, standard:1.15.0 should be runnable with this postgres template as well.

Reply:
I was thinking about that when was coding. I think that in addition to image types `postgres:nightly, istio:<ver> , standard:<ver>, wasm:<ver>` we should have a metadata - kind of softlink. So for example, postgres could point to standard:1.15.0, mysql could point to standard: 1.14.0. We could easily start adding flavors and keep number of builds fairly small.

Reply:
I also had https://github.com/tetratelabs/getenvoy/issues/57 issue in mind. Template repo which I added in this PR allows for default parameters, so not everything has to be specified in cmd line unless user wants to overwrite it.

Reply:
Yeah I think #57 is a good way to go.

> I was thinking about that when was coding. I think that in addition to image types `postgres:nightly, istio:<ver> , standard:<ver>, wasm:<ver>` we should have a metadata - kind of softlink. So for example, postgres could point to standard:1.15.0, mysql could point to standard: 1.14.0. We could easily start adding flavors and keep number of builds fairly small.

recent version of Envoy (1.13+) already reports its extension capability in admin endpoint (config_dump, https://github.com/envoyproxy/envoy/blob/master/api/envoy/config/core/v3/base.proto#L186), so in addition to #57 we can filter based on it to see if it has enough capability.